### PR TITLE
Sprite3D animation

### PR DESCRIPTION
**Pull request by super626**
Sprite3D Animation

Testcase in the Sprite3D Test
MAC/IOS, android, Windows, Linux have been tested. Can any one help to test other platforms

Features
3D animation Spport for Sprite3D.
New Sprite3D model file .c3t
fbx to c3t converter is available on Trace's repo.

### CODE REVIEW

File: cocos/2d/CCActionInterval.cpp [hunk 1]
```diff
@@ -528,6 +528,8 @@ void RepeatForever::step(float dt)
     if (_innerAction->isDone())
     {
         float diff = _innerAction->getElapsed() - _innerAction->getDuration();
+        if (diff > _innerAction->getDuration())
+            diff = fmodf(diff, _innerAction->getDuration());
```
Comment (ricardoquesada) [inline cocos/2d/CCActionInterval.cpp:532]: Why this change ?
What happens if you don't apply these 2 lines ?

Comment (dabingnn) [inline cocos/2d/CCActionInterval.cpp:532]: Here is the problem fixed in this PR https://github.com/cocos2d/cocos2d-x/pull/7086.
If there is a big lag, the animation will stop at the end of the animation.

**Comment by dabingnn** [comment]
Linux works now

**Comment by super626** [comment]
windows works

**Comment by super626** [comment]
windows works

**Comment by super626** [comment]
windows works

File: cocos/3d/CCAnimate3D.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    virtual void update(float t) override;
+    
+    float getSpeed() const { return _speed; }
+    void setSpeed(float speed) { _speed = speed; }
+    
+    bool getPlayBack() const { return _playBack; }
+    void setPlayBack(bool playBack) { _playBack = playBack; }
+protected:
+    Animate3D();
+    virtual ~Animate3D();
```
Comment (minggo) [inline cocos/3d/CCAnimate3D.h:70]: Please use `CC_CONSTRUCTOR_ACCESS` instead for constructor and destructor.

File: cocos/3d/CCAnimate3D.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+NS_CC_BEGIN
+
+//create Animate3D using Animation.
+Animate3D* Animate3D::create(Animation3D* animation)
+{
+    auto animate = new Animate3D();
+    animate->_animation = animation;
+    animation->retain();
+    
+    animate->autorelease();
```
Comment (minggo) [inline cocos/3d/CCAnimate3D.cpp:42]: Use retain and autorelease together.
So can remove these two lines?

Comment (minggo) [inline cocos/3d/CCAnimate3D.cpp:42]: Or, forget me.
animation and animate are different.

File: cocos/3d/CCAnimate3D.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    
+    return animate;
+}
+
+/** returns a clone of action */
+Animate3D* Animate3D::clone() const
+{
+    
+    auto animate = const_cast<Animate3D*>(this);
+    auto copy = animate->create(animate->_animation);
```
Comment (minggo) [inline cocos/3d/CCAnimate3D.cpp]: Why not just invoke Animate3D::create() instead?

File: cocos/3d/CCAnimation3D.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    
+    
+    float             _duration; //animation duration
+};
+
+class Animation3DCache
+{
+public:
+    static Animation3DCache* getInstance();
+    static void purgeMeshCache();
```
Comment (minggo) [inline cocos/3d/CCAnimation3D.h]: destroyInstance?

File: cocos/3d/CCAnimation3D.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    
+    Animation3D* getAnimation(const std::string& key);
+    
+    void addAnimation(const std::string& key, Animation3D* animation);
+    
+    void removeAllAnimation();
```
Comment (minggo) [inline cocos/3d/CCAnimation3D.h]: removeAllAnimation ->  removeAllAnimations

File: cocos/3d/CCAnimation3D.cpp [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+Animation3D* Animation3D::getOrCreate(const std::string& fileName, const std::string& animationName)
+{
+    std::string fullPath = FileUtils::getInstance()->fullPathForFilename(fileName);
+    std::string key = fullPath + "#" + animationName;
+    auto animation = Animation3DCache::getInstance()->getAnimation(key);
+    if (animation != nullptr)
+        return animation;
+    
+    //load animation here
+    animation = new Animation3D();
```
Comment (minggo) [inline cocos/3d/CCAnimation3D.cpp:42]: Should invoke autorelease?

File: cocos/3d/CCAnimationCurve.h [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+NS_CC_BEGIN
+
+enum EvaluateType
+{
+    Linear,
+    Near,
+    QuatSlerp,
+    UserFunction,
+};
```
Comment (minggo) [inline cocos/3d/CCAnimationCurve.h:44]: Use enum class instead and put it into class AnimationCurve?
And should use capital characters.
For example:
Linear -> LINEAR

Comment (minggo) [inline cocos/3d/CCAnimationCurve.h:44]: And use `_` to connect words, for example
QuatSlerp -> QUAT_SLERP

File: cocos/3d/CCBundle3D.h [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+NS_CC_BEGIN
+
+class Animation3D;
+class Bundle3D
+{
+public:
+    
+    static Bundle3D* getInstance();
+    
+    static void purgeBundle3D();
```
Comment (minggo) [inline cocos/3d/CCBundle3D.h]: destroyInstance?

File: cocos/3d/CCBundle3D.cpp [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+    const rapidjson::Value& bone_array_0 = skin_data_1["children"][(rapidjson::SizeType)0];
+    skindata->rootBoneIndex = skindata->getBoneNameIndex(bone_array_0["id"].GetString());
+    getChildMap(skindata->boneChild, skindata, bone_array_0);
+    return true;
+}
+
+/**
+ * load material data from bundle
+ * @param id The ID of the material, load the first Material in the bundle if it is empty
+ */
```
Comment (minggo) [inline cocos/3d/CCBundle3D.cpp]: I think we should add these comments in header files, then we can use doxygen to generate documents.

File: cocos/3d/CCMesh.h [hunk 11]
```diff
@@ -57,8 +49,12 @@ class RenderMeshData
     }
     bool hasVertexAttrib(int attrib);
     bool initFrom(const std::vector<float>& positions, const std::vector<float>& normals, const std::vector<float>& texs, const std::vector<unsigned short>& indices);
+    bool initFrom(const std::vector<float>& vertices, int vertexSizeInFloat, const std::vector<unsigned short>& indices, int numIndex, const std::vector<MeshVertexAttrib>& attribs, int attribCount);
```
Comment (minggo) [inline cocos/3d/CCMesh.h]: The api name is objective-c style. Initialize from what?
May be it is better to just name it init or add more clear description, initFromVertices.

File: plugin [hunk 12]
```diff
@@ -1 +1 @@
-Subproject commit 51ceded726cc034cdc939e379e1b24249e0ebad4
+Subproject commit 4a8c94a85a08cddbea5e1bc499c50cce7518fb15
```
Comment (minggo) [inline plugin]: Don't change submodule.

File: tools/cocos2d-console [hunk 13]
```diff
@@ -1 +1 @@
-Subproject commit f5037bab73a8fb109e8e34656220bed1a1743087
+Subproject commit 498f24c1683e4725ecaad6168a1aab21b283b8d5
```
Comment (minggo) [inline tools/cocos2d-console]: Should update submodule after synchronizing upstream code,
`git submodule update`

**Comment by minggo** [comment]
And please use smaller picture for test case. It is just to test the codes, not for demo. We have spent much energy to reduce the repo size.

File: cocos/3d/CCBundle3D.cpp (hunk 10 above)
Comment (super626) [inline cocos/3d/CCBundle3D.cpp]: I will change it according your comments a few minutes. Now, I am doing a
binary converter for mac.

**Comment by minggo** [comment]
[ci rebuild]

**Comment by minggo** [comment]
Oh.
You didn't replace the big picture.
It's my careless to merge it.

**Comment by super626** [comment]
Sorry, it is my fault. I didn't inform you that I don't have a photoshop. I asked Harrison to reduce the size of picture but without luck he did not receive the message in time.
